File: algo/week01/875-koko-eating-bananas/solution.py
# 875. 爱吃香蕉的珂珂
# https://leetcode.cn/problems/koko-eating-bananas/
import math


class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        lo, hi = 1, max(piles)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            hrs = sum(math.ceil(p / mid) for p in piles)
            if hrs <= h:
                hi = mid
            else:
                lo = mid + 1
        return lo

File: algo/week01/875-koko-eating-bananas/solution.ts
/**
 * 875. 爱吃香蕉的珂珂
 * https://leetcode.cn/problems/koko-eating-bananas/
 *
 * 二分速度 k：每小时吃 k 根，一堆吃完才能下一堆。
 */

export function minEatingSpeed(piles: number[], h: number): number {
  let lo = 1;
  let hi = Math.max(...piles);
  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);
    let hrs = 0;
    for (const p of piles) {
      hrs += Math.ceil(p / mid);
    }
    if (hrs <= h) hi = mid;
    else lo = mid + 1;
  }
  return lo;
}
